File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
const dy = [1, -1, 0, 0];
const dx = [0, 0, 1, -1];

function solution(maps) {
    const N = maps.length;
    const M = maps[0].length;
    const dist = Array.from({length: N}, ()=> Array.from({length: M}, ()=> 0));
    const queue = [[0, 0]];
    dist[0][0] = 1;
    
    while(queue.length>0){
        const [y, x] = queue.shift();
        
        if(y===N-1 && x===M-1){
            return dist[y][x];
        }
        
        for(let i=0; i<4; i++){
            const ny = y + dy[i];
            const nx = x + dx[i];
            
            if(ny<0||ny>=N||nx<0||nx>=M||dist[ny][nx]!==0||maps[ny][nx]===0)    continue;
            dist[ny][nx] = dist[y][x] + 1;
            queue.push([ny, nx]);
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dy = [1,-1,0,-0]
dx = [0,0,1,-1]

def solution(maps):
    N = len(maps)
    M = len(maps[0])
    
    queue = deque()
    visited = [[False]*M for _ in range(N)]
        
    queue.append((0, 0, 1))
    visited[0][0] = True
    
    while(queue):
        y, x, count = queue.popleft()
        
        if y == N-1 and x == M-1:
            return count
        
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            
            if 0 <= ny < N and 0 <= nx < M and not visited[ny][nx] and maps[ny][nx] == 1:
                visited[ny][nx] = True
                queue.append((ny, nx, count + 1))
    
    
    return -1